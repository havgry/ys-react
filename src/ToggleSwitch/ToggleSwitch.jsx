import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import uniqueId from 'lodash.uniqueid'

const id = `toggle-switch-${uniqueId()}`

const ToggleSwitch = ({
  label,
  defaultChecked,
  onChange,
  className,
  ...props
}) => {
  const wrapperProps = {
    className: classNames(
      'ys-toggle-switch',
      className,
    ),
    ...props,
  }

  return (
    <div
      {...wrapperProps}
    >
      <input
        type="checkbox"
        id={id}
        defaultChecked={defaultChecked}
        onChange={onChange}
        className="ys-toggle-switch__control ys-u-visuallyhidden"
      />
      <label
        htmlFor={id}
        className="ys-toggle-switch__label"
      >
        {label}
      </label>
    </div>
  )
}

ToggleSwitch.propTypes = {
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  defaultChecked: PropTypes.bool,
  className: PropTypes.string,
}

ToggleSwitch.defaultProps = {
  defaultChecked: false,
  className: '',
}

export default ToggleSwitch
