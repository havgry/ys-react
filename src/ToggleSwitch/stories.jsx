import React from 'react'
import { storiesOf } from '@storybook/react'

import { text, boolean } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import ToggleSwitch from './ToggleSwitch'
import '@youseedk/dna/css/elements/ys-toggle-switch.css'

storiesOf('ToggleSwitch', module)
  .addParameters({ jest: ['ToggleSwitch'] })
  .add('Default', () => (
    <ToggleSwitch
      onChange={action('onChange')}
      defaultChecked={boolean('defaultChecked', false, 'Content')}
      label={text('label', 'Label text', 'Content')}
      className={text('className', '', 'Misc')}
    />
  ))
