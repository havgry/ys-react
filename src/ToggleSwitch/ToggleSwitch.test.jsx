import React from 'react'
import { shallow, mount } from 'enzyme'

import ToggleSwitch from './ToggleSwitch'

describe('<ToggleSwitch>', () => {
  describe('Props', () => {
    it('Has correct defaultProps', () => {
      const component = mount(
        <ToggleSwitch
          onChange={jest.fn()}
          label="Click me!"
        />,
      )

      const {
        defaultChecked,
        className,
      } = component.props()

      expect(defaultChecked).toBe(false)
      expect(className).toBe('')
    })

    it('Throws propType warning if no `onChange` prop is passed', () => {
      expect(
        () => <ToggleSwitch
          label="Click me!"
        />,
      ).toThrow()
    })

    it('Throws propType warning if no `label` prop is passed', () => {
      expect(
        () => <ToggleSwitch
          onChange={jest.fn()}
        />,
      ).toThrow()
    })
  })

  describe('Rendering', () => {
    it('Outputs elements with the correct class names', () => {
      expect(
        shallow(
          <ToggleSwitch
            label="Click me!"
          />,
        )
          .find('div.ys-toggle-switch')
          .exists(),
      ).toBe(true)
    })

    it('Custom `className` is being applied to the container', () => {
      expect(
        shallow(
          <ToggleSwitch
            onChange={jest.fn()}
            label="Click me!"
            className="snail"
          />,
        )
          .find('.ys-toggle-switch.snail')
          .exists(),
      ).toBe(true)
    })

    it('Sets `label`', () => {
      expect(
        shallow(
          <ToggleSwitch
            onChange={jest.fn()}
            label="Toggle flying reptiles"
          />,
        )
          .find('.ys-toggle-switch__label')
          .text(),
      ).toBe('Toggle flying reptiles')
    })

    it('Forwards other generic props to container element', () => {
      expect(
        shallow(
          <ToggleSwitch
            onChange={jest.fn()}
            label="Click me!"
            id="polyphyly"
          />,
        ).props().id,
      ).toBe('polyphyly')
    })

    it('Forwards `onChange` and `defaltChecked` to input element', () => {
      const changeHandler = jest.fn()
      const component = shallow(
        <ToggleSwitch
          onChange={changeHandler}
          defaultChecked={true}
          label="Click me!"
        />,
      )

      const {
        onChange,
        defaultChecked,
      } = component
        .find('.ys-toggle-switch__control')
        .props()

      expect(onChange).toBe(changeHandler)
      expect(defaultChecked).toBe(true)
    })
  })

  describe('Snapshots', () => {
    it('With custom `className`', () => {
      expect(
        shallow(
          <ToggleSwitch
            onChange={jest.fn()}
            label="Toggle flying reptiles"
            className="pterosaur"
          />,
        ),
      ).toMatchSnapshot()
    })
  })
})
