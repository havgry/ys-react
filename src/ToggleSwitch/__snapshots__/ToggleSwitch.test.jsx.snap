// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleSwitch> Snapshots With custom \`className\` 1`] = `
<div
  className="ys-toggle-switch pterosaur"
>
  <input
    className="ys-toggle-switch__control ys-u-visuallyhidden"
    defaultChecked={false}
    id="toggle-switch-1"
    onChange={[MockFunction]}
    type="checkbox"
  />
  <label
    className="ys-toggle-switch__label"
    htmlFor="toggle-switch-1"
  >
    Toggle flying reptiles
  </label>
</div>
`;
